src/map_matrix.c

```diff
@@ -0,0 +1,231 @@
+#include "map_matrix.h"
+#include "safari_zone.h"
+#include <nitro/rtc.h>
+
+struct ScriptState;
+
+extern void* SavArray_Flags_get(SAVEDATA* savedata);
+extern void sub_02066C1C(struct ScriptState* state, u32);
+extern void sub_02066C4C(struct ScriptState* state, u32);
+extern BOOL CheckFlagInArray(struct ScriptState* state, u16 flag_id);
+extern void GF_RTC_CopyDate(RTCDate* date);
+extern SAFARIZONE* sub_0202F57C(SAVEDATA* savedata);
+extern SAFARIZONE_UNKSUB1* sub_0202F630(SAFARIZONE* safari_zone, s32);
+
+static void MapMatrix_MapMatrixData_Load(MAPMATRIXDATA* map_matrix_data, u16 matrix_id, u32 map_no) {
+    map_matrix_data->width = 0;
+    map_matrix_data->height = 0;
+
+    s32 i;
+
+    for (i = 0; i < MAP_MATRIX_MAX_SIZE; i++) {
+        map_matrix_data->headers[i] = 0;
+        map_matrix_data->altitudes[i] = 0;
+        map_matrix_data->maps.data[i] = 0;
+    }
+
+    for (i = 0; i < MAP_MATRIX_MAX_NAME_LENGTH; i++) {
+        map_matrix_data->name[i] = 0;
+    }
+
+    void* buffer = AllocAtEndAndReadWholeNarcMemberByIdPair(NARC_fielddata_mapmatrix_map_matrix, matrix_id, 11);
+    u8* cursor = (u8*)buffer;
+
+    map_matrix_data->width = *(cursor++);
+    map_matrix_data->height = *(cursor++);
+
+    u8 has_headers_section = *(cursor++);
+    u8 has_altitudes_section = *(cursor++);
+
+    u8 name_length = *(cursor++);
+    GF_ASSERT(name_length <= MAP_MATRIX_MAX_NAME_LENGTH);
+
+    MI_CpuCopy8(cursor, &map_matrix_data->name, name_length);
+    cursor += name_length;
+
+    if (has_headers_section) {
+        MI_CpuCopy8(cursor, &map_matrix_data->headers, map_matrix_data->width * map_matrix_data->height * sizeof(u16));
+        cursor += map_matrix_data->width * map_matrix_data->height * sizeof(u16);
+    } else {
+        MIi_CpuClear16((u16)map_no, (u16*)&map_matrix_data->headers, map_matrix_data->width * map_matrix_data->height * sizeof(u16));
+    }
+
+    if (has_altitudes_section) {
+        MI_CpuCopy8(cursor, &map_matrix_data->altitudes, map_matrix_data->width * map_matrix_data->height * sizeof(u8));
+        cursor += map_matrix_data->width * map_matrix_data->height * sizeof(u8);
+    }
+
+    MI_CpuCopy8(cursor, map_matrix_data->maps.data, map_matrix_data->width * map_matrix_data->height * sizeof(u16));
+    FreeToHeap(buffer);
+}
+
+MAPMATRIX* MapMatrix_New(void) {
+    MAPMATRIX* map_matrix = AllocFromHeap(11, sizeof(MAPMATRIX));
+    map_matrix->width = 0;
+    map_matrix->height = 0;
+    map_matrix->matrix_id = 0;
+
+    return map_matrix;
+}
+
+void MapMatrix_Load(u32 map_no, MAPMATRIX* map_matrix) {
+    u16 matrix_id = MapHeader_GetMatrixId(map_no);
+
+    MapMatrix_MapMatrixData_Load(&map_matrix->data, matrix_id, map_no);
+
+    map_matrix->matrix_id = matrix_id;
+    map_matrix->height = map_matrix->data.height;
+    map_matrix->width = map_matrix->data.width;
+}
+
+void MapMatrix_Free(MAPMATRIX* map_matrix) {
+    FreeToHeap(map_matrix);
+}
+
+u16 MapMatrix_GetMapData(s32 map_no, MAPMATRIX* map_matrix) {
```

Comment:
What kind of data is this? Tile metadata? Map ID at selected offset?

Reply:
It gets the map model number at a selected offset. I'll update the names.